miner/worker.go

```diff
@@ -1076,16 +1081,89 @@ func (w *worker) prepareWork(genParams *generateParams) (*environment, error) {
 // into the given sealing block. The transaction selection and ordering strategy can
 // be customized with the plugin in the future.
 func (w *worker) fillTransactions(interrupt *int32, env *environment) error {
-	// Split the pending transactions into locals and remotes
-	// Fill the block with all available pending transactions.
+	// First pull Electroneum Ltd transactions out of 'pending' if flag --prioritiseElectroneum was passed by
+	// cli argument and then Split the pending transactions into locals and remotes
+	// pending is a local copy of the pool not a pointer, so it's ok to strip out Electroneum tx from localTxs and
+	// remoteTxs before processing them separately and avoid double counting.
 	pending := w.eth.TxPool().Pending(true)
+
+	electroneumTxs := make(map[common.Address]types.Transactions)
+	if w.config.PrioritiseElectroneum {
+
+		//todo: move this into worker or somewhere to save doing this every time we fill a block
+		//testing keys only
+		// private key : 8dd88ba44afa275e0dd9be1d633732242d84ea955e453b28719d17897035b4c2
+		//NB the 0x04 is to tell us that the key is uncompressed 64bytes +1 byte
+		etnSigningPubKey := "04" + "d6e7d6da18e8cfb9cf20bbb7182ac36deb9b77f2c617f3698fc12aebf22ee9b00060775349a2312813818a9c7a0a4b143fddfa36611129daf01c26bc9080d418"
+		byteStringETNKey, err := hex.DecodeString(etnSigningPubKey)
+		if err != nil {
+			fmt.Println("Error decoding hex string:", err)
+			panic("Error decoding etn signing key")
+		}
+
+		// If the first transaction for a given account has the Electroneum Ltd signature present, then all of the tx
+		// for that account do, so we can just pull that account out of pending without further thought
+		for account, txs := range pending {
+			// Extract the nonce - it's a uint64, so we need to convert it to bytes
+			nonceHex := hexutil.EncodeUint64(txs[0].Nonce())
+			nonce, _ := hex.DecodeString(nonceHex[2:]) // [2:] to trim the '0x' prefix
+			// Create signer depending on whether the transaction is protected
+			var signer types.Signer
+			if txs[0].Protected() {
+				signer = types.NewEIP155Signer(w.chainConfig.ChainID)
+			} else {
+				signer = types.HomesteadSigner{}
```

Comment:
the `signer` can be defined outside the for loop, ie. having `signer` and `eip155Signer` instantiated outside the for-loop and used accordingly depending if tx is protected or not.

Reply:
signer is a per transaction thing, so it's inside the for-loop which loops over all pending tx

Reply:
I can't see why the signer is a per transaction thing, you instantiate them by either:
- `signer = types.NewEIP155Signer(w.chainConfig.ChainID)`
- `signer = types.HomesteadSigner{}`

No transaction reference during signer instantiation here.

But I agree that either regular signer or eip155 signer should be used based on transaction type. What I suggested was to instantiate both at a higher level (outside the for-loop) and use either `signer` or `eip155signer` in the for-loop (where it is now) depending if tx is protected or not.

Reply:
I just meant that the type of signer depended on the the tx being assessed. Can definitely be declared outside though. I think i'd just left it in for readability because it's a cheap operation, but it need not be inside. However going to just remove the signer logic altogether because the tx is verified earlier at the txpool level & the sanity check on 'sender' is a little excessive